hw/ip/otbn/util/check_loop.py

```diff
@@ -0,0 +1,226 @@
+    '''
+    out = CheckResult()
+    for l in loops:
+        insn = program.get_insn(l.start)
+        operands = program.get_operands(l.start)
+        if insn.mnemonic == 'loopi' and operands['iterations'] <= 0:
+            out.err(
+                'Bad number of loop iterations ({}) at PC {:#x}: {}'.format(
+                    operands['iterations'], l.start,
+                    insn.disassemble(l.start, operands)))
+    return out
+
+
+def _check_loop_end_insns(program: OTBNProgram,
+                          loops: List[CodeSection]) -> CheckResult:
+    '''Checks that loops do not end in control flow instructions.
+
+    Such instructions can cause LOOP software errors during execution.
+    '''
+    out = CheckResult()
+    for l in loops:
+        loop_end_insn = program.get_insn(l.end)
+        if not loop_end_insn.straight_line:
+            out.err(
+                'Control flow instruction ({}) at end of loop at PC {:#x} (loop starting at PC {:#x})'
+                .format(loop_end_insn.mnemonic, l.end, l.start))
+    return out
+
+
+def _check_loop_branching(program: OTBNProgram,
+                          loops: List[CodeSection]) -> CheckResult:
+    '''Checks that there are no branches into or out of loop bodies.
+
+    Branches within the same loop body are permitted (but not branches from an
+    inner loop to an outer loop, as this counts as branching out of the inner
+    loop). Because this isn't necessarily a fatal issue (for instance, it's
+    possible the branched-to code will always return to the loop), this check
+    returns warnings rather than errors.
+
+    A `jal` instruction with a register other than `x1` as the first operand is
+    treated the same as a branch and not permitted to cross the loop-body
+    boundary.
+    '''
+    out = CheckResult()
+
+    # Check all bne and beq instructions, as well as `jal` instructions with
+    # first operands other than x1 (unconditional branch)
+    to_check = _get_branches(program)
+    for pc in _get_pcs_for_mnemonic(program, 'jal'):
+        operands = program.get_operands(pc)
+        if operands['grd'] != 1:
+            to_check.append(pc)
+
+    for pc in to_check:
+        operands = program.get_operands(pc)
+        branch_addr = operands['offset'] & ((1 << 32) - 1)
+
+        # Get the loop bodies the branch is inside, if any
+        current_loops = []
+        for l in loops:
+            if pc in l:
+                current_loops.append(l)
+
+        # Check that we're not branching out of any loop bodies
+        for l in current_loops:
+            if branch_addr not in l:
+                insn = program.get_insn(pc)
+                out.warn(
+                    'Branch out of loop at PC {:#x} (loop from PC {:#x} to PC {:#x}, branch {} to PC {:#x}). This might cause problems with the loop stack and surprising behavior.'
+                    .format(pc, l.start, l.end, insn.mnemonic, branch_addr))
```

Comment:
Oof! I don't like the long line here. Fortunately, Python does the same string concatenation thing as C, so you can write this as e.g.
```python3
                out.warn('Branch out of loop at PC {:#x} (loop from PC {:#x} '
                         'to PC {:#x}, branch {} to PC {:#x}). This might '
                         'cause problems with the loop stack and surprising '
                         'behavior.'
                         .format(pc, l.start, l.end,
                                 insn.mnemonic, branch_addr))
```

Reply:
Hmm, `util/lintpy.py --fix` keeps changing this back for me and putting the format on the same line.

Reply:
Yeah, our Python auto-formatter makes a mess of nice code. Fortunately, it's not mandated by our style guide. No strong feelings here: happy to leave it ugly if it's hard to find a formulation that the tool doesn't mess up.